application-load-balancer/lambda_function/secret_rotator.py

```diff
@@ -0,0 +1,624 @@
+                        'HeaderValue': pending_secret['HEADERVALUE']
+                    }]
+                    logger.info(f"Custom header added in CloudFront distribution: {origin['Id']}")
+                    # Mark that we modified this distribution by adding the header
+                    all_have_header = False
+                else:
+                    # If 'Items' exists, check if the custom header is present
+                    header_found = any(
+                        header['HeaderName'] == self.header_name
+                        for header in origin['CustomHeaders']['Items']
+                    )
+
+                    # If the header is not found, add it
+                    if not header_found:
+                        logger.info(f"Custom header not found in origin {origin['Id']}, adding header.")
+                        origin['CustomHeaders']['Items'].append({
+                            'HeaderName': self.header_name,
+                            'HeaderValue': pending_secret['HEADERVALUE']
+                        })
+
+                        logger.info(f"Custom header found/added in CloudFront distribution: {origin['Id']}")
+                        all_have_header = False  # Mark this as needing update, since we added it
+
+                # If header is found in the Items, we can break out of the loop for this origin
+                if header_found:
+                    break
+
+            # If header was not found and added in any of the origins, we mark all_have_header as False
+            if not header_found:
+                all_have_header = False
+
+        if all_have_header:
+            # If all CF distributions have the header, update WAF rule first
+            logger.info("Updating WAF rule first. All CloudFront distributions already have custom header.")
+            self.update_waf_acl(pending_secret['HEADERVALUE'], current_secret['HEADERVALUE'])
+
+            # Sleep for 75 seconds for regional WAF config propagation
+            logger.info("Sleeping for 75 seconds for updated WAF rule propagation.")
+            time.sleep(75)
+
+        # Update each CloudFront distribution
+        for distro in matching_distributions:
+            try:
+                logger.info(f"Updating CloudFront distribution {distro['Id']}.")
+                self.update_cf_distro(distro['Id'], pending_secret['HEADERVALUE'])
+            except Exception as e:
+                logger.error(f"Failed to update distribution {distro['Id']}: {e}")
+                raise
+
+        if not all_have_header:
+            # If not all CF distributions had the header, update WAF last
+            logger.info("Not all CloudFront distributions have the header. Updating WAF last.")
+            self.update_waf_acl(pending_secret['HEADERVALUE'], current_secret['HEADERVALUE'])
+
+            # Sleep for 75 seconds for regional WAF config propagation
+            logger.info("Sleeping for 75 seconds for WAF rule propagation.")
```

Comment:
Also are there any tests that hit these areas of code? (here and lines 296-302)

If there are, then this will result in *really* slow tests without a way of overriding the number of secs. If there aren't, then they'll need writing (along with a way of overriding this value)

Reply:
You would have to write a method that calls:
`aws secretsmanager describe-secret \
            --secret-id "$SECRET_ID" \
            --query 'LastRotatedDate' \
            --output text`
 and checks that the LastRotatedDate value has changed. I do something similar in the e2e test

Reply:
I've added tests that test the order of precedence in which `update_waf_acl` and `update_cf_distro` are called depending on the initial state of the CF distribution

Reply:
This has been added as a variable.  However we are mocking that in the test so we wont get that wait.